src/riak_ensemble_router.erl

```
@@ -0,0 +1,252 @@
+%%% gen_server callbacks
+%%%===================================================================
+
+init([]) ->
+    {ok, #state{}}.
+
+handle_call(_Request, _From, State) ->
+    {reply, ok, State}.
+
+handle_cast({ensemble_cast, Ensemble, Msg}, State) ->
+    ensemble_cast(Ensemble, Msg),
+    {noreply, State};
+handle_cast(_Msg, State) ->
+    {noreply, State}.
+
+handle_info(_Info, State) ->
+    {noreply, State}.
+
+terminate(_Reason, _State) ->
+    ok.
+
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+
+-spec ensemble_cast(ensemble_id(), msg()) -> error | ok.
+ensemble_cast(Ensemble, Msg) ->
+    case riak_ensemble_manager:get_leader(Ensemble) of
+        {_, Node}=Leader ->
+            %% io:format("L: ~p~n", [Leader]),
+            if Node =:= node() ->
+                    Pid = riak_ensemble_manager:get_peer_pid(Ensemble, Leader),
+                    %% io:format("Sending to ~p~n", [Pid]),
+                    handle_ensemble_cast(Msg, Pid),
+                    ok;
+               true ->
+                    riak_ensemble_router:cast(Node, Ensemble, Msg),
+                    ok
+            end;
+        undefined ->
+            error
+    end.
+
+-spec handle_ensemble_cast(_,_) -> ok.
+handle_ensemble_cast({sync_send_event, From, Ref, Event, Timeout}, Pid) ->
+    spawn(fun() ->
+                  try
+                      Result = gen_fsm:sync_send_event(Pid, Event, Timeout),
+                      From ! {Ref, Result}
+                  catch
+                      _:_ ->
+                          From ! {Ref, timeout}
+                  end
+          end),
+    ok;
+handle_ensemble_cast(_, _Pid) ->
+    ok.
+
+fail_cast({sync_send_event, From, Ref, _Event, _Timeout}) ->
+    From ! {Ref, timeout};
+fail_cast(_) ->
```

AFAICT this function only gets called from one place with the form above. This catchall clause seems like a good place to fail fast instead